adal/tests/Test.ADAL.NET.Unit/DeviceCodeFlowTests.cs

```diff
@@ -93,5 +93,71 @@ public async Task PositiveTestAsync()
             Assert.IsNotNull(result);
             Assert.AreEqual("some-access-token", result.AccessToken);
         }
+
+        [TestMethod]
+        public async Task AdfsPositiveTestAsync()
+        {
+            DeviceCodeResult dcr = new DeviceCodeResult()
+            {
+                ClientId = TestConstants.DefaultClientId,
+                Resource = TestConstants.DefaultResource,
+                DeviceCode = "device-code",
+                ExpiresOn = (DateTimeOffset.UtcNow + TimeSpan.FromMinutes(10)),
+                Interval = 5,
+                Message = "get token here",
+                UserCode = "user-code",
+                VerificationUrl = "https://contoso.com/adfs/oauth2/deviceauth"
+            };
+
+            AdalHttpMessageHandlerFactory.InitializeMockProvider();
+            MockHttpMessageHandler mockMessageHandler = new MockHttpMessageHandler()
+            {
+                Method = HttpMethod.Post,
+                Url = "https://login.contoso.com/adfs/oauth2/token",
+                ResponseMessage = MockHelpers.CreateFailureResponseMessage("{\"error\":\"authorization_pending\"," +
+                                                                           "\"error_description\":\"AADSTS70016: Pending end-user authorization." +
+                                                                           "\\r\\nTrace ID: f6c2c73f-a21d-474e-a71f-d8b121a58205\\r\\nCorrelation ID: " +
+                                                                           "36fe3e82-442f-4418-b9f4-9f4b9295831d\\r\\nTimestamp: 2015-09-24 19:51:51Z\"," +
+                                                                           "\"error_codes\":[70016],\"timestamp\":\"2015-09-24 19:51:51Z\",\"trace_id\":" +
+                                                                           "\"f6c2c73f-a21d-474e-a71f-d8b121a58205\",\"correlation_id\":" +
+                                                                           "\"36fe3e82-442f-4418-b9f4-9f4b9295831d\"}")
+            };
+
+            AdalHttpMessageHandlerFactory.AddMockHandler(mockMessageHandler);
+            AdalHttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler()
+            {
+                Method = HttpMethod.Post,
+                Url = "https://login.contoso.com/adfs/oauth2/token",
+                ResponseMessage =
+                    MockHelpers.CreateSuccessTokenResponseMessage(TestConstants.DefaultUniqueId,
+                        TestConstants.DefaultDisplayableId, TestConstants.DefaultResource)
+            });
+
+            TokenCache cache = new TokenCache();
+            AuthenticationContext ctx = new AuthenticationContext(TestConstants.DefaultAdfsAuthorityTenant, false, cache);
+            AuthenticationResult result = await ctx.AcquireTokenByDeviceCodeAsync(dcr);
+            Assert.IsNotNull(result);
+            Assert.AreEqual("some-access-token", result.AccessToken);
```

Comment:
You can add an extra check to these tests by asserting that all the Http mock handlers were consumed. We do this in many of our other tests, I think we missed a bug at some point in the past by not doing it.